### PR TITLE
Ensure state updater is notified on workflow completion.

### PR DESCRIPTION
**Pull request by nishkrishnan**
The following is how this looks when there is an error.  Eventually we can parse out Terraform specific errors but it's not necessary right now.

image

### CODE REVIEW

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 1]
```diff
@@ -243,9 +263,9 @@ func (r *Runner) Run(ctx workflow.Context) error {
 		return r.toExternalError(err, "fetching root")
 	}
 	defer func() {
-		err := cleanup()
+		cleanupErr := cleanup()
 
-		if err != nil {
+		if cleanupErr != nil {
 			logger.Warn(ctx, "error cleaning up local root", "err", err)
```
Comment (Aayyush) [inline server/neptune/workflows/internal/terraform/workflow.go:269]: nit: should this be `cleanupErr`?  
```suggestion
			logger.Warn(ctx, "error cleaning up local root", "err", cleanupErr)
```

File: server/neptune/workflows/internal/deploy/terraform/state.go [hunk 2]
```diff
@@ -124,22 +118,13 @@ func (n *StateReceiver) emitApplyEvents(ctx workflow.Context, jobState *state.Jo
 }
 
 func determineCheckRunState(workflowState *state.Workflow) github.CheckRunState {
```
Comment (Aayyush) [inline server/neptune/workflows/internal/deploy/terraform/state.go:120]: thoughts on using a switch statement here?

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 3]
```diff
@@ -221,6 +221,26 @@ func (r *Runner) Apply(ctx workflow.Context, root *terraform.LocalRoot, serverUR
 func (r *Runner) Run(ctx workflow.Context) error {
 	r.MetricsHandler.Gauge(ActiveTerraformWorkflowStat).Update(1)
 	defer r.MetricsHandler.Gauge(ActiveTerraformWorkflowStat).Update(0)
+	var err error
+	// make sure we are updating state on completion.
+	defer func() {
+		reason := state.SuccessfulCompletionReason
+		if r := recover(); r != nil || err != nil {
```
Comment (samrabelachew) [inline server/neptune/workflows/internal/terraform/workflow.go:228]: when will err not be nil

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 4]
```diff
@@ -221,6 +221,26 @@ func (r *Runner) Apply(ctx workflow.Context, root *terraform.LocalRoot, serverUR
...
+			reason = state.InternalServiceError
```
Comment (Aayyush) [inline server/neptune/workflows/internal/terraform/workflow.go:229]: If I'm understanding correctly, if we see any error in the workflow run (`PlanRejectedError` for eg), we set the reason to `state.InternalServiceError`?

File: server/neptune/workflows/internal/terraform/workflow.go (hunk 3 above)
Comment (nishkrishnan) [reply server/neptune/workflows/internal/terraform/workflow.go:228]: what do you mean? err is set to the value of `run`?

File: server/neptune/workflows/internal/terraform/workflow.go (hunk 4 above)
Comment (nishkrishnan) [reply server/neptune/workflows/internal/terraform/workflow.go:229]: In the future we'd want to show custom messages based on the error, but given the time crunch right now I'd rather just fix the problem at hand and move on right now.

Comment (Aayyush) [reply server/neptune/workflows/internal/terraform/workflow.go:229]: Sounds good to me! Just wanted to make sure

File: server/neptune/workflows/internal/terraform/workflow.go (hunk 3 above)
Comment (samrabelachew) [reply server/neptune/workflows/internal/terraform/workflow.go:228]: oops missed that

**Review by Aayyush — APPROVED** [review]

**Review by samrabelachew — APPROVED** [review]

**Pull request merged**